### DISCUSSION TITLE
I want to execute an op only after all partitions have finished materializing for my job

### DISCUSSION THREAD

**Discussion by magoocas** [dagster-io/dagster#18676] [Q&A] [answered]
Not sure if this is possible.

I want to execute an op only after all partitions have finished materializing for my job.

Or perhaps to put it another way, I have a partitioned job that I need to trigger an unpartitioned job only when all the partitions are complete. Currently my unpartitioned job is being executed every time the partitioned job finishes a partition.

**Comment by sryza** [comment, marked as answer]
@magoocas are these modeled as assets? If so, you can select the assets and launch a backfill, and unpartitioned assets will wait until  all of the targeted upstream partitions are materialized before being executed.

**Reply by magoocas** [reply]
Yes they are assets, however I am trying to materialize them all using a job. In my testing the down stream unpartitioned asset is being materialized every time an upstream partition finishes.

**Reply by magoocas** [reply]
@sryza I've created a toy example of how I have things setup:


``` python
from pprint import pformat
from typing import Dict
from dagster import (
    AssetExecutionContext,
    Definitions,
    DynamicPartitionsDefinition,
    DailyPartitionsDefinition,
    MultiPartitionsDefinition,
    OpExecutionContext,
    Output,
    asset,
    op,
    job,
    define_asset_job,
)

element_partitions_def = DynamicPartitionsDefinition(name="element_partitions")

daily_partitions_def = DailyPartitionsDefinition(
    start_date="2023-12-10", end_date="2023-12-12", timezone="Etc/GMT+7"
)

daily_elements_def = MultiPartitionsDefinition(
    {
        "element": element_partitions_def,
        "date": daily_partitions_def,
    }
)


@op
def download_data_partitions(context: OpExecutionContext):
    context.log.info("Generating data partitions")
    partitions = ["A", "B", "C"]

    context.instance.add_dynamic_partitions("element_partitions", partitions)


@asset(
    partitions_def=daily_elements_def,
)
def upstream_partitioned(context: AssetExecutionContext) -> Output[str]:
    context.log.info(f"Processing partition {context.partition_key}")
    return Output(str(context.partition_key))


@asset
def downstream_unpartitioned(
    context: AssetExecutionContext, upstream_partitioned: Dict[str, str]
) -> Output[str]:
    context.log.info(f"Processing partition {context.partition_key}")
    context.log.info(
        "Received upstream partitioned asset:\n" + pformat(upstream_partitioned)
    )
    return Output(str(context.partition_key))


@job
def download_data_partitions_job():
    download_data_partitions()


materialize_assets_job = define_asset_job(
    "materialize_assets",
    [upstream_partitioned, downstream_unpartitioned],
    partitions_def=daily_elements_def,
)

defs = Definitions(
    assets=[upstream_partitioned, downstream_unpartitioned],
    jobs=[download_data_partitions_job, materialize_assets_job],
)
```
No matter how I try to materialize this, whether through the ui or via a job as shown a backfill for all partitions causes (in this case) `downstream_unpartitioned` to be materialized 6 times.

**Reply by sryza** [reply]
Ah, got it. This is because job backfills don't yet respect partition mappings: https://github.com/dagster-io/dagster/issues/11962.

If you launch the backfill directly from the asset graph, it should work correctly.

**Comment by parkerdaminer** [comment]
Maybe try telling Dagster you want your backfill run in 1 run. Otherwise it will try to run in separate jobs to be faster. I've not tested this code, but I use this parameter in some of my jobs where I need it all run in a single backfill operation.

```python
@asset(
    partitions_def=daily_elements_def,
    backfill_policy=BackfillPolicy.single_run(),
)
def upstream_partitioned(context: AssetExecutionContext) -> Output[str]:
    context.log.info(f"Processing partition {context.partition_key}")
    return Output(str(context.partition_key))
```

**Reply by magoocas** [reply]
Interesting, I've messed around with that before. Trying it now, I can't even get the upstream asset to materialize in a single run. 

I think there is something going on with the MultiPartitionsDefinition. I must be doing something wrong there.

Even this doesn't materialize as expected:

``` python
element_partitions_def = DynamicPartitionsDefinition(name="element_partitions")

daily_partitions_def = DailyPartitionsDefinition(
    start_date="2023-12-10", end_date="2023-12-12", timezone="Etc/GMT+7"
)

daily_elements_def = MultiPartitionsDefinition(
    {
        "element": element_partitions_def,
        "date": daily_partitions_def,
    }
)


@op
def download_data_partitions(context: OpExecutionContext):
    context.log.info("Generating data partitions")
    partitions = ["A", "B", "C"]

    context.instance.add_dynamic_partitions("element_partitions", partitions)


@asset(
    partitions_def=daily_elements_def,
    backfill_policy=BackfillPolicy.single_run(),
)
def upstream_partitioned(context: AssetExecutionContext) -> Output[str]:
    context.log.info(f"Processing partition {context.partition_key}")
    return Output(str(context.partition_key))


@job
def download_data_partitions_job():
    download_data_partitions()


materialize_assets_job = define_asset_job(
    "materialize_assets",
    [upstream_partitioned],
    partitions_def=daily_elements_def,
)

defs = Definitions(
    assets=[upstream_partitioned],
    jobs=[download_data_partitions_job, materialize_assets_job],
)
```

**Reply by magoocas** [reply]
@parkerdaminer do you have a working example of using BackfillPolicy.single_run()?

I'm going crazy trying to get this to work in any case.

**Reply by magoocas** [reply]
Nevermind the docs explicitly mention this isn't possible. 

[Launching single-run backfills using backfill policies ](https://docs.dagster.io/concepts/partitions-schedules-sensors/backfills#launching-single-run-backfills-using-backfill-policies-)

(See the note)

**Reply by parkerdaminer** [reply]
I've got this working with a non MultiPartitionDefinition, I'm using Daily Partition for my asset. Not sure if it works for Multi.

**Reply by j-blackwell** [reply]
@parkerdaminer, have you had any joy with this? I am not sure it is currently possible to use a single run backfill policy on a multi partition. I am able to return it from my function, but dagster gets stuck in the `output` stage. Specifically:


`DbIOManager._get_table_slice`
``` python
...
if isinstance(context.asset_partitions_def, MultiPartitionsDefinition):
    multi_partition_key_mapping = cast(
        MultiPartitionKey, context.asset_partition_key
    ).keys_by_dimension
...
```

But there is no `asset_partition_key` as it is a partition range instead:

```
dagster._check.CheckError: Failure condition: Tried to access partition key for output 'result' of step 'my_asset', but the step output has a partition range: 'A|2015-01-01' to 'A|2023-12-01'.
```

I am pretty sure this is a limitation, currently.

**Reply by j-blackwell** [reply]
#19408
